File: Gstar/continuous/kernelD.m
%
% Function: kernelD(input)
%
% outputs the 2n*ns dimensional vector DK(H)(w)
% approximates dK/dHj
%
% Input: H = substituted CRS,
%        w = n*1 vector contains frequencies,
%        s = relaxation modes,
%
% Output: DK = Jacobian of H
%

function DK = kernelD(H, w, s)

	ns         = length(s);
	hs         = zeros(ns,1);
	hs(1)      = 0.5*log(s(2)/s(1));
	hs(ns)     = 0.5*log(s(ns)/s(ns-1));
	hs(2:ns-1) = 0.5 * (log(s(3:ns))-log(s(1:ns-2)));

	n  = length(w);

	[S, W]  = meshgrid(s,w);
	ws      = S.*W;
	clear S W;

	ws2     = ws.^2;
	Hsuper  = ones(2*n,1)*(hs' .* exp(H'));       % A 2n*ns matrix with all the rows = H'

	DK      = [ (ws2./(1+ws2)) ;(ws./(1+ws2)) ] .* Hsuper;

end

File: ReSpect_v2/Gt/kernelD.r
# Function: kernelD
#
# outputs the n*ns dimensional vector DK(H)(t)
# approximates dK/dHj
#
# Input: H = substituted CRS,
#        t = n*1 vector of times,
#        s = relaxation modes,
#
# Output: DK = Jacobian of H
#

kernelD <- function(H, t, s) {
  
  ns <- length(s)
  hs <- numeric(ns)
  hs[1] <- 0.5 * log(s[2] / s[1])
  hs[ns] <- 0.5 * log(s[ns] / s[ns-1])
  hs[2:(ns-1)] <- 0.5 * (log(s[3:ns]) - log(s[1:(ns-2)]))
  
  n <- length(t)
  
  kern <- outer(t, s, function(T, S) exp(-T / S))
  
  Hsuper <- matrix(rep(hs * exp(H), each = n), nrow = n, ncol = ns, byrow = TRUE)
  
  DK <- kern * Hsuper
  
  return(DK)
}
